geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/jobs/PartitionOutputFormat.scala

```diff
@@ -51,49 +54,135 @@ class PartitionOutputFormat(delegate: SingleFileOutputFormat) extends OutputForm
+      val metadata = StorageMetadataFactory.load(fsc).getOrElse {
         throw new IllegalArgumentException(s"No storage defined under path '$root'")
       }
+      FileSystemStorageFactory(fsc, metadata)
     }
+    private val encoding = storage.metadata.encoding
+    private val leaf = storage.metadata.leafStorage
+
     private val fileType = StorageConfiguration.getFileType(context.getConfiguration)
+    private val fileSize = StorageConfiguration.getTargetFileSize(context.getConfiguration)
 
     private val counter = context.getCounter(Group, Features)
     private val cache = scala.collection.mutable.Map.empty[String, PartitionState]
 
-    override def write(key: Void, value: SimpleFeature): Unit = {
-      val partition = metadata.scheme.getPartitionName(value)
-      val state = cache.getOrElseUpdate(partition, createWriter(partition))
-      state.writer.write(key, value)
-      val geom = value.getDefaultGeometry.asInstanceOf[Geometry]
-      if (geom != null) {
-        state.bounds.expandToInclude(geom.getEnvelopeInternal)
+    private val workPath = delegate.getOutputCommitter(context).asInstanceOf[FileOutputCommitter].getWorkPath
+
+    private def newState(partition: String): PartitionState = {
+      val estimator = storage match {
+        case s: SizeableFileSystemStorage => s.targetSize(fileSize).map(s.estimator)
+        case _ => None
       }
-      state.count += 1L
+      estimator match {
+        case None => new SinglePartitionState(partition)
+        case Some(e) => new ChunkedPartitionState(partition, e, context)
+      }
+    }
+
+    override def write(key: Void, value: SimpleFeature): Unit = {
+      val partition = storage.metadata.scheme.getPartitionName(value)
+      val state = cache.getOrElseUpdate(partition, newState(partition))
+      state.write(key, value)
       counter.increment(1)
     }
 
     override def close(context: TaskAttemptContext): Unit = {
       cache.foreach { case (partition, state) =>
         logger.debug(s"Closing writer for $partition")
-        state.writer.close(context)
-        val files = Seq(StorageFile(state.file, System.currentTimeMillis()))
-        metadata.addPartition(PartitionMetadata(partition, files, PartitionBounds(state.bounds), state.count))
+        state.close(context)
+        storage.metadata.addPartition(state.meta)
+      }
+      CloseWithLogging(storage)
+    }
+
+    sealed abstract class PartitionState(partition: String) {
+
+      private var count = 0L
+      private val bounds = new Envelope()
+      private val files = ArrayBuffer.empty[String]
+
+      def write(key: Void, value: SimpleFeature): Unit = {
+        val geom = value.getDefaultGeometry.asInstanceOf[Geometry]
+        if (geom != null) {
+          bounds.expandToInclude(geom.getEnvelopeInternal)
+        }
+        count += 1L
+      }
+
+      def meta: PartitionMetadata = {
+        val f = files.map(StorageFile(_, System.currentTimeMillis()))
```

Comment:
Do we care about calling `System.currentTimeMills()` repeatedly?  (It seems like it could return multiple different times if the operation spans more than a millisecond.

Reply:
it's ok if the times aren't the same, but yeah seems like we could save a few calls. I'll update.